### PR TITLE
W-11117613: Duplicate application of lifecycle phases in Global Error…

### PR DESCRIPTION
**Pull request by marianorean**
… Handlers

### CODE REVIEW

File: integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java [hunk 1]
```diff
@@ -49,6 +51,14 @@ protected String getConfigFile() {
   @Named("flowD")
   private FlowConstruct flowD;
 
+  @Inject
```
Comment (fsgonz) [inline integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java:54]: Add also tests indicating what the correct behavior for the start and initialise phase should be in the case of a GlobalErrorHandler.

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java [hunk 2]
```diff
@@ -49,6 +51,14 @@ protected String getConfigFile() {
...
+  @Named("flowE")
+  private FlowConstruct flowE;
+
+  @Inject
+  @Named("flowF")
+  private FlowConstruct flowF;
+
   @Test
   public void testLifecycleErrorHandlerInFlow() throws Exception {
```
Comment (fsgonz) [inline integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java:63]: verify in this tests also that there are no double application of phases. I think you can set a default property of the error tracker for this.

**Review by fsgonz — CHANGES_REQUESTED** [review]

**Review by fsgonz — DISMISSED** [review]

**Review by fsgonz — DISMISSED** [review]

File: integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import javax.inject.Inject;
+import javax.inject.Named;
+
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.Rule;
+
+public class DefaultErrorHandlerLifecycleTestCase extends ErrorHandlerLifecycleTestCase {
```
Comment (fsgonz) [inline integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java:34]: Add @allure

**Review by fsgonz — CHANGES_REQUESTED** [review]

File: integration/src/test/resources/org/mule/test/integration/exceptions/default-error-handler-lifecycle.xml [hunk 4]
```diff
@@ -10,21 +10,33 @@
...
     <error-handler name="defaultEh">
         <on-error-continue>
-            <test-components:lifecycle-tracker name="defaultEhErrorHandlerTracker"/>
+            <test-components:lifecycle-tracker-check name="defaultEhErrorHandlerTracker"/>
+        </on-error-continue>
+    </error-handler>
+
+    <error-handler name="global">
+        <on-error-continue>
+            <test-components:lifecycle-tracker name="globalErrorHandlerTracker"/>
```
Comment (elrodro83) [inline integration/src/test/resources/org/mule/test/integration/exceptions/default-error-handler-lifecycle.xml:31]: what is the difference between `lifecycle-tracker` and `lifecycle-tracker-check`? whis is this one nor check?

File: integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java (hunk 3 above)
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java:34]: split (or parameterize) this test do it explicitely tests global or default error handlers

Comment (elrodro83) [reply integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java:34]: same applies for the other test class changed in this PR

File: integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java [hunk 5]
```diff
@@ -9,8 +9,10 @@
 import static org.hamcrest.core.Is.is;
 import static org.junit.Assert.assertThat;
 
+import org.mule.runtime.api.lifecycle.Disposable;
 import org.mule.runtime.api.lifecycle.Initialisable;
 import org.mule.runtime.api.lifecycle.Lifecycle;
+import org.mule.runtime.api.lifecycle.Startable;
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java:15]: add allure annotations to this class as well

**Review by elrodro83 — CHANGES_REQUESTED** [review]

File: integration/src/test/resources/org/mule/test/integration/exceptions/default-error-handler-lifecycle.xml (hunk 4 above)
Comment (marianorean) [reply integration/src/test/resources/org/mule/test/integration/exceptions/default-error-handler-lifecycle.xml:31]: The `lifecycle-tracker-check` checks that the lifecycle phases are called only once, so it is necessary to use `lifecycle-tracker` in a test that starts, stops and starts again a flow.

File: integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+  }
+
+  @Test
+  public void testLifecycleDefaultErrorHandler() throws Exception {
+    flowRunner(flowD.getName()).run();
+
+    Collection<String> defaultEhErrorHandlerPhases = trackersRegistry.get("defaultEhErrorHandlerTracker").getCalledPhases();
+
+    assertThat(defaultEhErrorHandlerPhases.contains(Initialisable.PHASE_NAME), is(true));
+    assertThat(defaultEhErrorHandlerPhases.contains(Startable.PHASE_NAME), is(true));
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/integration/exceptions/DefaultErrorHandlerLifecycleTestCase.java:87]: use collection matchers

**Review by elrodro83 — COMMENTED** [review]
I put the comments just once but they apply to the 3 test cases involved in this PR

File: integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java [hunk 7]
```diff
@@ -87,10 +99,12 @@ public void testLifecycleDefaultErrorHandler() throws Exception {
     Collection<String> defaultEhErrorHandlerPhases = trackersRegistry.get("defaultEhErrorHandlerTracker").getCalledPhases();
 
     assertThat(defaultEhErrorHandlerPhases.contains(Initialisable.PHASE_NAME), is(true));
+    assertThat(defaultEhErrorHandlerPhases.contains(Startable.PHASE_NAME), is(true));
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java:102]: collection matchers

File: integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java [hunk 8]
```diff
@@ -87,10 +99,12 @@ public void testLifecycleDefaultErrorHandler() throws Exception {
...
 
     ((Lifecycle) flowD).stop();
+    ((Lifecycle) flowD).dispose();
 
     assertThat(defaultEhErrorHandlerPhases.contains(Stoppable.PHASE_NAME), is(true));
+    assertThat(defaultEhErrorHandlerPhases.contains(Disposable.PHASE_NAME), is(true));
```
Comment (elrodro83) [inline integration/src/test/java/org/mule/test/integration/exceptions/ErrorHandlerLifecycleTestCase.java:108]: collection matchers

**Review by elrodro83 — CHANGES_REQUESTED** [review]

**Review by fsgonz — APPROVED** [review]

**Pull request merged**